endless/src/endless/EndlessUsbToolDlg.cpp

```diff
@@ -5131,6 +5134,31 @@ bool CEndlessUsbToolDlg::CopyFilesToESP(const CString &fromFolder, const CString
 	return retResult;
 }
 
+bool CEndlessUsbToolDlg::CreatePersistentStorageFileOnexFAT(const CString& drive)
+{
+    HANDLE tmpfile;
+    CString storage = drive + L"\\endless\\persistent.img";
+    char marker[] = "endless_live_storage_marker";
+    uint64_t bytes;
+    ULARGE_INTEGER free_bytes;
+    DWORD written;
+
+    if (!GetDiskFreeSpaceEx(drive, &free_bytes, NULL, NULL))
+        return FALSE;
+
+    bytes = free_bytes.QuadPart;
+    tmpfile = CreateFile(storage, GENERIC_WRITE, FILE_SHARE_WRITE, NULL, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL);
+    if (tmpfile == INVALID_HANDLE_VALUE)
+        return FALSE;
+
+    // Ignore error because we're going to keep going anyway.
+    WriteFile(tmpfile, marker, strlen(marker), &written, NULL);
+
+    safe_closehandle(tmpfile);
+
+    return ExtendImageFile(storage, free_bytes.QuadPart);
```

Comment:
IIRC `ExtendImageFile` is cheap because it avoids zeroing out the data it allocates, so it reveals whatever happened to be on the disk before. For `C:\endless\endless.img` this seems totally OK because the file is only legible by the administrator. Here we're taking the majority of a USB stick and making its contents visible to anyone who can read the disk, i.e. all users on the system.

I was about to say this is an issue because I might create a USB stick and then give it to my friend. However, it was already true that my friend could read the raw bytes back off the disk in that case.

So, it's probably OK.

Reply:
Right, that's exactly how ExtendImageFile works.

Similarly, the current dd'd iso image live usb sticks leak info, even after the reformat step, as that only writes to some of the formatted space.

If you're paranoid, go back in time and encrypt the stuff you were storing on the usb stick before you repurposed it.

I don't think it's worthwhile to turn making usb sticks into an all day affair for this...
We could add a checkbox to zero the device? But even then, the dd image approach would require manual zeroing of the device.

Reply:
I think it's OK, I was just thinking out loud.

Reply:
But – the current ISO stuff doesn't leak info to non-root users because it just creates a partition on the uninitialized space, whereas this is creating a world-readable file containing uninitialized space.